### PR TITLE
[FLINK-35776] Simplify job status handling

### PR DESCRIPTION
**Pull request by gyfora**
## What is the purpose of the change

There is a fairly complicated listing / observe logic for jobs currently that is no longer necessary as we have a stable logic to always record the jobID in the status before submission.

## Brief change log

 - Replace complex resource specific job listing/matching logic with JobId based querying
 - Clean up unnecessary codepaths
 - Tests

## Verifying this change

The change is already covered by existing tests to a large extent.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: yes

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Review by ferenc-csaky — COMMENTED** [review]
Added 1 nitty comment, the REST LGTM, although I'm not particularly familiar with this code path, hence the lack of approve :)

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/AbstractFlinkService.java [hunk 1]
```diff
@@ -263,14 +268,24 @@ protected SocketAddress getSocketAddress(RestClusterClient<String> clusterClient
...
+                    .thenApply(
+                            mjd -> {
+                                if (mjd.getJobs() == null) {
+                                    return Optional.<JobStatusMessage>empty();
+                                }
+                                return mjd.getJobs().stream()
+                                        .filter(jd -> jd.getJobId().equals(jobId))
+                                        .findAny()
+                                        .map(JobStatusUtils::toJobStatusMessage);
+                            })
```
[the remark this answers is not in the document]
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/AbstractFlinkService.java:288]: I tried this @ferenc-csaky however the spotless line length limit breaks the lines a way that the chained optional is basically unreadable. I would prefer to keep it this way for readability.

**Comment by afedulov** [comment]
@gyfora 
> as we have a stable logic to always record the jobID in the status before submission.
could you point to the related changes to understand the context a bit better?

**Review by ferenc-csaky — APPROVED** [review]
LGTM, but needs conflict resolution now.

**Pull request merged**